File: 0050-powx-n/0050-powx-n.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        @cache
        def calc(a, p):
            if p == 0: return 1
            if p == 1: return a
            if p == -1: return 1/a

            if p % 2:
                return a * calc(a, p // 2) * calc(a, p // 2)
            return calc(a, p // 2) * calc(a, p // 2)


        return calc(x, n)

File: 0050-powx-n/0050-powx-n.java
class Solution {
    public double myPow(double x, long n) {
        if (n == 0) {
            return 1;
        }
        if (n < 0) {
            return myPow(1/x, Math.abs(n));
        }

        if (n % 2 == 1) {
            return x * myPow(x * x, n / 2);
        }
        return myPow(x * x, n / 2);
    }
}